src/Swift.Runtime/src/Swift/Runtime/ProtocolWitnessTable.cs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+
+namespace Swift.Runtime;
+
+public readonly struct ProtocolConformanceDescriptor : IEquatable<ProtocolConformanceDescriptor>
+{
+    private readonly IntPtr _handle;
+
+    private ProtocolConformanceDescriptor(IntPtr handle)
+    {
+        _handle = handle;
+    }
+
+    public readonly static ProtocolConformanceDescriptor Zero = default;
+
+    public bool IsValid => _handle != IntPtr.Zero;
+
+    public bool Equals(ProtocolConformanceDescriptor other)
+    {
+        return _handle == other._handle;
+    }
+
+    internal static bool TryGetProtocolConformanceDescriptor<T, U>([NotNullWhen(true)] out ProtocolConformanceDescriptor? result)
+    {
+        var type = typeof(T);
+
+        if (typeof(ISwiftObject).IsAssignableFrom(type))
+        {
+            var helperType = typeof(ProtocolConformanceDescriptorHelper<,>).MakeGenericType(typeof(T), typeof(U));
+            result = (ProtocolConformanceDescriptor)helperType.GetMethod("GetProtocolConformanceDescriptor")!.Invoke(null, null)!;
+            return true;
+        }
+
+        result = null;
+        return false;
+    }
+
+    public static ProtocolConformanceDescriptor LoadProtocolConformanceDescriptor(string libraryPath, string symbolName)
```

Comment:
Where are we going to get the `libraryPath` from?
There are potential security implications when loading code from file paths... so we should be careful.
If this is supposed to be just like PInvokes, then we should use the mechanism from the other overload of `NativeLibrary.Load`.

Reply:
So the `libraryPath` will be always known at compile time (this is the path of the dylib the type is declared in). What would be the safest way to obtain it using the symbol? I used `NativeLibrary.Load` instead of `PInvoke` as the` Protocol Conformance Descriptor` is not a function within the dylib as opposed to the `Metadata Accessor`.

Reply:
We can't rely on the path being the same between compile time and runtime. Even system libraries can be in different locations potentially (in the future) and non-system ones are in effectively random locations - for example if it ships with the app.
I think this should rely on the same resolution mechanism as PInvoke (unless I'm missing something). I understand we can't use PInvoke directly, but we should be able to use NativeLibrary.Load(name, assembly, ...) instead.

Reply:
@jkurdek Do we plan to replace it with the MacO reader? If there is no accessor function, can we read the content of the descriptor?

Reply:
@vitek-karas you are right, I meant library name. I will follow the pattern you outlined. 

@kotlarmilos Correct me if I'm wrong. My understanding is that during compile time we will use `MachO` to obtain the symbol for the protocol conformance descriptor. However, obtaining the protocol conformance descriptor needs to happen during runtime (as we need to use the platform version of the library). And in order to read that symbol we need to load native library right?

Reply:
> @kotlarmilos Correct me if I'm wrong. My understanding is that during compile time we will use MachO to obtain the symbol for the protocol conformance descriptor. However, obtaining the protocol conformance descriptor needs to happen during runtime (as we need to use the platform version of the library). And in order to read that symbol we need to load native library right?

Yes, that is right.